docs/src/main/asciidoc/security-authorize-web-endpoints-reference.adoc

```diff
@@ -386,6 +386,204 @@ It is possible to use multiple expressions in the role definition.
+        return "modified";
+    }
+
+    @PermissionsAllowed(value = {"create", "update"}, inclusive=true) <2>
+    @POST
+    public String createOrUpdate(Long id) {
+        return id + " modified";
+    }
+
+    @PermissionsAllowed({"see:detail", "see:all", "read"}) <3>
+    @GET
+    @Path("/id/{id}")
+    public String getItem(String id) {
+        return "item-detail-" + id;
+    }
+
+    @PermissionsAllowed(value = "permission-1", permission = CustomPermission.class) <4>
+    @GET
+    public Collection<String> list(@QueryParam("query-options") String queryOptions) {
+        // your business logic comes here
+        return Collections.emptySet();
+    }
+
+    public static class CustomPermission extends Permission {
+
+        public CustomPermission(String name) {
+            super(name);
+        }
+
+        @Override
+        public boolean implies(Permission permission) {
+            var event = Arc.container().instance(RoutingContext.class).get(); <5>
+            return "public-content".equals(event.request().params().get("query-options"));
+        }
+
+        ...
+    }
+}
+----
+<1> Resource method `createOrUpdate` is only accessible by user with both `create` and `update` permissions.
+<2> By default, at least one of the permissions specified through one annotation instance is required.
+You can require all of them by setting `inclusive=true`. Both resource methods `createOrUpdate` has equal authorization requirements.
+<3> Access is granted to `getItem` if `SecurityIdentity` has either `read` permission or `see` permission and one of actions (`all`, `detail`).
+<4> You can use any `java.security.Permission` implementation of your choice.
+By default, string-based permission is performed by the `io.quarkus.security.StringPermission`.
+<5> Permissions are not beans, therefore only way to obtain bean instances is programmatically via the `Arc.container()`.
+
+CAUTION: If you plan to use the `@PermissionsAllowed` on the IO thread, review the information in xref:security-proactive-authentication-concept.adoc[Proactive Authentication].
+NOTE: The `@PermissionsAllowed` is not repeatable on class-level due to limitations of Quarkus interceptors.
+Please find well-argued explanation in the xref:cdi-reference.adoc#repeatable-interceptor-bindings[Repeatable interceptor bindings] section of the CDI reference.
+
+On CDI beans, it is also possible to take custom `java.security.Permission` much further.
+Let's image that in your application, you have following service:
+
+[source,java]
+----
+import io.quarkus.security.PermissionsAllowed;
+import io.smallrye.mutiny.Uni;
+import jakarta.enterprise.context.ApplicationScoped;
+
+import java.security.Permission;
+import java.util.Objects;
+
+@ApplicationScoped
+public class ProductOrderService {
+
+    @PermissionsAllowed(value = "order:member-only", permission = OrderPermission.class) <1>
+    public Uni<String> getProduct(CustomerDetail customerDetail, OrderDetail orderDetail) {
```

Comment:
@michalvavrik It is an interesting example which shows how an extra `OrderDetail` parameter can complement the permission check, but there is no association of this permission with the authenticated identity.

I do see it now how this CDI PermissionAllowed idea works and I agree it can be useful and quite cool, but I don't see any relationship to the security access control process. It is an application level decision making. Any authenticated user can access this method so how can you restrict it to a member-only order detail specification in this case ?

IMHO it has to be wired a bit differently. Permission should be added as you say below (the only way currently is via `SecurityIdentityAugmentor`, roles as well as OIDC scopes etc support will follow) but this Permission, instead of requiring it to have an extra constructor parameter, is required to have a method `void set(OrderDetails details) `, which will be called by the time `OrderDetails` is available but before `implies(Permission)` is called. 
This way, in `SecurityIdentityAugmentor` (or via future mapings) one can decide if it is a member only request or not.

Reply:
Hmm, I still believe the concept is right and it's totally relevant what arguments with which method was invoked are useful for security access control process. 

I'll try to reflect what you say, but permission can't have extra methods, it has to be immutable.Let my try again with a new example before we switch a concept from constructors. Also **I'll add some explanations to docs, as I can see it is less intuitive for users than I thought!**

Few points @sberyozkin (but it's my mistake, better documentation and examples are needed, misjudged situation):

- I made mistakes where I used `OrderDetails`, I should use it inside `implies` as there, incoming `Permission` is coming from the `SecurityIdentityAugmentor`
- `SecurityIdentituAugmentor` won't ever be able to set `set(OrderDetails details)` as it has no idea what actual method arguments are. The augmentor creates users permissions, while `OrderDetails details` is created when method is invoked. Then flow is **at least one of permissions created augmentor must imply permission created for method invocation context (the one if order details)**
- I suggest instead of cutting out capabilities, I'll simplify examples and this extra stuff is for users to explore when needed

Reply:
@michalvavrik As I said, any authenticated user can have this method invoked, so there is no extra access control with this permission unless some mapping is introduced.

> SecurityIdentituAugmentor won't ever be able to set set(OrderDetails details) as it has no idea what actual method arguments are.

I did not propose it, sorry. What I said that `SecurityIdentituAugmentor` or other mapping mechanism (using some request path properties check, or role mapping, etc), associates an `order:admin` Permission with the current identity. Next, when enforcing the PermissionAllowed restriction, the security code, gets `order:admin` Permission from the identity and finds a `set(OrderDetails)` method on this Permission instance (if needed an extra interface like `PermissionProperty<T> { void set(T) }` can be added), before the actual check.

Reply:
@michalvavrik Let me also clarify, letting users do application level authorization decisions (ex order data is not the ones which are expected - return 403, etc) with the annotation is interesting, I agree, but if we don't care about the permission name/action while checking for the custom property then it should not be done as part of the PermissionsAllowed process, since the users would needs to extends Permission class add some random name in the annotation, follow some conventions (with the constructor or extra method).
So IMHO for this extra property check be done via PermissionAllowed the permission has to be mapped to the identity (starting with the augmentor for now), if it does not matter what the authenticated identity has for this permission be checked then it just should be an independent annotation not part of Quarkus Security (ParameterPermission or something like that, part of Resteasy reactive) which would require a custom class has a single constructor parameter only. 
We can both options supported really, but in scope of this PR please analyze if the permission with an extra check can be bound to the identity as proposed
thanks

Reply:
We've agreed with @michalvavrik that the docs will need to be updated a bit, what I'm trying to suggest here was already thought about by Michal earlier